sunkit_image/granule.py

```diff
@@ -0,0 +1,298 @@
+"""
+This module contains functions that will segment images for granule detection.
+"""
+import logging
+
+import numpy as np
+import scipy.ndimage as sndi
+import skimage
+from sklearn.cluster import KMeans
+
+import sunpy
+import sunpy.map
+
+__all__ = [
+    "segment",
+    "get_threshold",
+    "trim_intergranules",
+    "mark_brightpoint",
+    "kmeans_segment",
+    "correlation",
+]
+
+METHODS = ["li", "otsu", "isodata", "mean", "minimum", "yen", "triangle"]
+
+
+def segment(smap, resolution, *, skimage_method="li", mark_dim_centers=False, bp_min_flux=None):
+    """
+    Segment an optical image of the solar photosphere into tri-value maps with:
+
+     * 0 as intergranule
+     * 1 as granule
+     * 1.5 as brightpoint 

```

Comment:
Maybe there is a particular reason for this, but why 1.5 for a brightpoint? Why not 2? If you're data are all integers, this will make your resulting map much smaller (as opposed using 1.5 which means everything must be a float).

Reply:
The reasoning was simply that we thought it seemed cleaner. We considered using 0=intergranule, 1=dim_middle, 2=granule, 3=brightpoint, but having granules be the seeminlgy arbitrary value of 2 seemed odd. We also considered 0=intergranule, 1=granule, 2=brightpoint, 3=dim_middle, but we wanted the final fits file to "make sense" if plotted as-is, e.g. for dim_middles to be a lower value than granules, and brightpoints to be a higher value. However, if it significantly decreases the file size, that may outweigh having labels that seem inuitive.

Reply:
I like that mask -> intensity logic, but I think having masks of integer values makes more sense (and doesn't run into floating point equivalency issues) and can save space (but compressing an array of identical floats is pretty efficient). Trying to maintain a brightness progression in the segment maps is just going to run into trouble if someone wants to add a fifth type of feature in the future and is not the primary use of the mask maps.

So I would suggest this scheme proposed above: 0=intergranule, 1=granule, 2=brightpoint, 3=dim_middle
Granules and intergranules, the two most clearly defined and understood features are at the beginning of the ordering scheme, with easily remembered values, and the other features follow

Reply:
To make this a bit less ambiguous, you could, when creating the output map, add a discrete colormap and appropriate norm that indicates these are discrete values. See [this example](https://matplotlib.org/stable/tutorials/colors/colormap-manipulation.html#sphx-glr-tutorials-colors-colormap-manipulation-py) for one way to do this. You can then add this to the `plot_settings` on the `Map` object so that any time `.plot()` is called on that Map, it uses that colormap and norm. [This example](https://docs.sunpy.org/en/stable/generated/gallery/plotting/map_editcolormap.html) should point you in the right direction.

Reply:
Ok, I can work on this!

Reply:
I like the approach of the map carrying a suitable colormap for a comprehensible display. But please, not too much of a rainbow!